sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala

```diff
@@ -2079,4 +2078,654 @@ case class GpuConv(num: Expression, fromBase: Expression, toBase: Expression)
+      val intPartPos = closeOnExcept(decTemp) { _ =>
+        withResource(intPart) { _ =>
+          withResource(Scalar.fromString("-")) { negativeSign =>
+            withResource(Scalar.fromString("")) { emptyString =>
+              intPart.stringReplace(negativeSign, emptyString)
+            }
+          }
+        }
+      }
+      resource_array += intPartPos
+      val appendZeros = "0" * appendZeroNum
+      val appendZerosCv = closeOnExcept(decTemp) { _ =>
+        withResource(Scalar.fromString(appendZeros)) { zeroString =>
+          ColumnVector.fromScalar(zeroString, cv.getRowCount.toInt)
+        }
+      }
+      val decPart = withResource(decTemp) { _ =>
+        withResource(appendZerosCv) { _ =>
+          ColumnVector.stringConcatenate(Array(decTemp, appendZerosCv))
+        }
+      }
+      (intPartPos, decPart)
+    }
+  }
+
+  private def getParts(cv: ColumnVector, d: Int): (ColumnVector, ColumnVector) = {
+    // get int part and dec part from a column vector, int part will be set to positive
+    x.dataType match {
+      case FloatType | DoubleType => {
+        getPartsFromDouble(cv, d)
+      }
+      case DecimalType.Fixed(_, scale) => {
+        getPartsFromDecimal(cv, d, scale)
+      }
+      case IntegerType | LongType | ShortType | ByteType => {
+        closeOnExcept(ArrayBuffer.empty[ColumnVector]) { resource_array =>
+          val intPartPos = withResource(cv.castTo(DType.STRING)) { intPart =>
+            withResource(Scalar.fromString("-")) { negativeSign =>
+              withResource(Scalar.fromString("")) { emptyString =>
+                intPart.stringReplace(negativeSign, emptyString)
+              }
+            }
+          }
+          resource_array += intPartPos
+          val dzeros = "0" * d
+          val decPart = withResource(Scalar.fromString(dzeros)) { zeroString =>
+            ColumnVector.fromScalar(zeroString, cv.getRowCount.toInt)
+          }
+          (intPartPos, decPart)
+        }
+      }
+      case _ => {
+        throw new UnsupportedOperationException(s"format_number doesn't support type ${x.dataType}")
+      }
+    }
+  }
+
+  private def negativeCheck(cv: ColumnVector): ColumnVector = {
+    withResource(cv.castTo(DType.STRING)) { cvStr =>
+      withResource(Scalar.fromString("-")) { negativeSign =>
+        cvStr.startsWith(negativeSign)
+      }
+    }
+  }
+
+  private def removeExtraCommas(str: ColumnVector): ColumnVector = {
+    withResource(Scalar.fromString(",")) { comma =>
+      str.rstrip(comma)
+    }
+  }
+
+  private def addCommas(str: ColumnVector): ColumnVector = {
+    val maxstrlen = withResource(str.getCharLengths()) { strlen =>
+      withResource(strlen.max()) { maxlen =>
+        maxlen.isValid match {
+          case true => maxlen.getInt
+          case false => 0
+        }
+      }
+    }
+    val substrs = (0 until maxstrlen by 3).map { i =>
+      val endIndex = i + 3
+      str.substring(i, endIndex).asInstanceOf[ColumnView]
```

Comment:
It is because array is invariant in scala. I will got:
```
[ERROR] [Error] /home/haoyangl/spark-rapids/sql-plugin/src/main/scala/org/apache/spark/sql/rapids/stringFunctions.scala:2583: type mismatch;
 found   : Array[ai.rapids.cudf.ColumnVector]
 required: Array[ai.rapids.cudf.ColumnView]
Note: ai.rapids.cudf.ColumnVector <: ai.rapids.cudf.ColumnView, but class Array is invariant in type T.
You may wish to investigate a wildcard type such as `_ <: ai.rapids.cudf.ColumnView`. (SLS 3.2.10)
```
if removing this.